### PR TITLE
ARROW-17610: [C++] Support additional source types in SourceNode

### PR DESCRIPTION
**Pull request by rtpsw**
See https://issues.apache.org/jira/browse/ARROW-17610

### CODE REVIEW

**Comment by rtpsw** [comment]
cc @westonpace, @pitrou

File: cpp/src/arrow/compute/exec/plan_test.cc [hunk 1]
```diff
@@ -296,6 +296,85 @@ TEST(ExecPlanExecution, TableSourceSinkError) {
               Raises(StatusCode::Invalid, HasSubstr("batch_size > 0")));
 }
 
+template <typename ElementType, typename OptionsType>
+void test_source_sink_error(
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/plan_test.cc:300]:
```suggestion
void TestSourceSinkError(
```

File: cpp/src/arrow/compute/exec/plan_test.cc [hunk 2]
```diff
@@ -296,6 +296,85 @@ TEST(ExecPlanExecution, TableSourceSinkError) {
...
+  auto null_executor_options = OptionsType{exp_batches.schema, element_it_maker};
+  ASSERT_OK(MakeExecNode(source_factory_name, plan.get(), {}, null_executor_options));
+
+  auto null_schema_options = OptionsType{no_schema, element_it_maker};
+  ASSERT_THAT(MakeExecNode(source_factory_name, plan.get(), {}, null_schema_options),
+              Raises(StatusCode::Invalid, HasSubstr("not null")));
+}
+
+template <typename ElementType, typename OptionsType>
+void test_source_sink(
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/plan_test.cc:322]:
```suggestion
void TestSourceSink(
```

File: cpp/src/arrow/compute/exec/plan_test.cc [hunk 3]
```diff
@@ -296,6 +296,85 @@ TEST(ExecPlanExecution, TableSourceSinkError) {
...
+TEST(ExecPlanExecution, RecordBatchSourceSink) {
+  test_source_sink<std::shared_ptr<RecordBatch>, RecordBatchSourceNodeOptions>(
+      "record_batch_source", ToRecordBatches);
+}
+
+TEST(ExecPlanExecution, RecordBatchSourceSinkError) {
+  test_source_sink_error<std::shared_ptr<RecordBatch>, RecordBatchSourceNodeOptions>(
+      "record_batch_source", ToRecordBatches);
+}
+
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/plan_test.cc:377]:
```suggestion
TEST(ExecPlanExecution, ArrayVectorSourceSink) {
  TestSourceSink<std::shared_ptr<ArrayVector>, ArrayVectorSourceNodeOptions>(
      "array_vector_source", ToArrayVectors);
}

TEST(ExecPlanExecution, ArrayVectorSourceSinkError) {
  TestSourceSinkError<std::shared_ptr<ArrayVector>, ArrayVectorSourceNodeOptions>(
      "array_vector_source", ToArrayVectors);
}

TEST(ExecPlanExecution, ExecBatchSourceSink) {
  TestSourceSink<std::shared_ptr<ExecBatch>, ExecBatchSourceNodeOptions>(
      "exec_batch_source", ToExecBatches);
}

TEST(ExecPlanExecution, ExecBatchSourceSinkError) {
  TestSourceSinkError<std::shared_ptr<ExecBatch>, ExecBatchSourceNodeOptions>(
      "exec_batch_source", ToExecBatches);
}

TEST(ExecPlanExecution, RecordBatchSourceSink) {
  TestSourceSink<std::shared_ptr<RecordBatch>, RecordBatchSourceNodeOptions>(
      "record_batch_source", ToRecordBatches);
}

TEST(ExecPlanExecution, RecordBatchSourceSinkError) {
  TestSourceSinkError<std::shared_ptr<RecordBatch>, RecordBatchSourceNodeOptions>(
      "record_batch_source", ToRecordBatches);
}

```

File: cpp/src/arrow/compute/exec/options.h [hunk 4]
```diff
@@ -77,6 +78,34 @@ class ARROW_EXPORT TableSourceNodeOptions : public ExecNodeOptions {
...
+/// \brief An extended Source node which accepts a schema
+///
+/// ItMaker is a maker of an iterator of tabular data.
+template <typename ItMaker>
+class ARROW_EXPORT SchemaSourceNodeOptions : public ExecNodeOptions {
+ public:
+  SchemaSourceNodeOptions(std::shared_ptr<Schema> schema, ItMaker it_maker)
+      : schema(schema), it_maker(std::move(it_maker)) {}
+
+  // the schema of the record batches from the iterator
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/options.h:90]:
```suggestion
  /// the schema of the record batches from the iterator
```

**Review by westonpace — CHANGES_REQUESTED** [review]
I still think there is value in the user being able to specify the I/O executor.  I just think we should default it to the default I/O executor.

File: cpp/src/arrow/compute/exec/options.h [hunk 5]
```diff
@@ -77,6 +78,34 @@ class ARROW_EXPORT TableSourceNodeOptions : public ExecNodeOptions {
   int64_t max_batch_size;
 };
 
+/// \brief An extended Source node which accepts a schema
+///
+/// ItMaker is a maker of an iterator of tabular data.
+template <typename ItMaker>
+class ARROW_EXPORT SchemaSourceNodeOptions : public ExecNodeOptions {
```
[the remark this answers is not in the document]
Comment (rtpsw) [reply cpp/src/arrow/compute/exec/options.h:85]: I'm not sure whether it is normal to expose an internal API, such as `arrow::internal::Executor`, in a public API such as here. I'll include it in the next commit, but please confirm this is acceptable.

**Comment by rtpsw** [comment]
@westonpace, any idea how to fix the [Windows linker error here](https://github.com/apache/arrow/actions/runs/3122413945/jobs/5064339466)? In the past I was able to fix this by adding `inline` but this time it didn't work.

**Comment by rtpsw** [comment]
> @westonpace, any idea how to fix the [Windows linker error here](https://github.com/apache/arrow/actions/runs/3122413945/jobs/5064339466)? In the past I was able to fix this by adding `inline` but this time it didn't work.

I suspect this is due to adding `Executor`, which is not a public symbol, in a constructor parameter as you requested. If so and we insist on exposing the IO executor, I see a couple of options to consider:
1. Make `Executor` a public symbol. I'm not sure what issues that might lead to.
2. Make `Executor` derive from a new type which would be a public symbol, and use this type in the constructor. This would lead to a not-so-pretty dynamic cast internally.
3. Expose one or more different (public-symbol) parameters that are used to internally setup the IO executor. This restricts the kind of IO executors that the caller may set up.

**Comment by westonpace** [comment]
`Executor` is exported and it is included in other places that are part of the public API (e.g. SourceNodeOptions, CSV reader).

I think the problem is that `SchemaSourceNodeOptions` is templated.  Can you try adding something like (note, the first two lines already exist in your PR)...

```
using ArrayVectorIteratorMaker = std::function<Iterator<std::shared_ptr<ArrayVector>>()>;
using ArrayVectorSourceNodeOptions = SchemaSourceNodeOptions<ArrayVectorIteratorMaker>;
template class ARROW_EXPORT ArrayVectorSourceNodeOptions; // Explicitly export the specialization
```

**Comment by rtpsw** [comment]
It turns out template instantiations and attributes (in this case, visibility) do not work together nicely. When I try adding code like `template<> class ARROW_EXPORT ArrayVectorSourceNodeOptions`, I get compiler errors like:
```
/mnt/user1/tscontract/github/rtpsw/arrow/cpp/src/arrow/compute/exec/options.h:113:31: error: using typedef-name ‘using ArrayVectorSourceNodeOptions = class arrow::compute::SchemaSourceNodeOptions<std::function<arrow::Iterator<std::shared_ptr<std::vector<std::shared_ptr<arrow::Array> > > >()> >’ after ‘class’
  113 | template<> class ARROW_EXPORT ArrayVectorSourceNodeOptions;
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
and when I try adding code like `template<> class ARROW_EXPORT SchemaSourceNodeOptions<ArrayVectorIteratorMaker>`, I get compiler warnings like:
```
/mnt/user1/tscontract/github/rtpsw/arrow/cpp/src/arrow/compute/exec/options.h:113:31: warning: attributes ignored on template instantiation [-Wattributes]
  113 | template<> class ARROW_EXPORT SchemaSourceNodeOptions<ArrayVectorIteratorMaker>;
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
So, both tries do not work, at least with my compiler. There seems to be a [proposal documenting this limitation](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0537r0.html), but I'm not sure what its status is.

In the commit I just made, I opted instead to explicitly define and export non-template symbols.

**Comment by rtpsw** [comment]
The CI failures look unrelated to the PR.

**Comment by rtpsw** [comment]
@westonpace, to make sure we're on the same page, what remains for this PR to go through?

**Review by westonpace — APPROVED** [review]
This looks good now and should be very useful for helping people get data into Acero, thanks!  I'll wait until after we make the RC branch (I think this happens on Monday) before I merge.

**Comment by rtpsw** [comment]
@westonpace, is it a good time to get back to this PR?

**Comment by westonpace** [comment]
Yes, apologies. Since it's been a month lets give it one last CI pass.  I've rebased this (all tests pass locally) and will merge later today.

**Pull request merged**